tests/test_qnn.py

```diff
@@ -0,0 +1,429 @@
+# limitations under the License.
+"""
+Tests for the pennylane.qnn module.
+"""
+import os
+import tempfile
+
+import numpy as np
+import pytest
+
+try:
+    import tensorflow as tf
+    from tensorflow.keras.layers import Layer
+    from tensorflow.keras.initializers import RandomNormal
+
+    CORRECT_TF_VERSION = int(tf.__version__.split(".")[0]) > 1
+except ImportError:
+    from abc import ABC
+
+    Layer = ABC
+    CORRECT_TF_VERSION = False
+
+import pennylane as qml
+from pennylane.qnn import KerasLayer
+
+min_tf = pytest.mark.skipif(
+    not CORRECT_TF_VERSION, reason="TensorFlow version 2 and above required"
+)
+
+
+@pytest.fixture
+def get_circuit(n_qubits, output_dim, interface):
+    """Fixture for getting a sample quantum circuit with a controllable qubit number and output
+    dimension. Returns both the circuit and the shape of the weights."""
+
+    dev = qml.device("default.qubit", wires=n_qubits)
+    weight_shapes = {"w1": (3, n_qubits, 3), "w2": (1,), "w3": 1, "w4": [3], "w5": (2, n_qubits, 3)}
+
+    @qml.qnode(dev, interface=interface)
+    def circuit(inputs, w1, w2, w3, w4, w5):
+        """A circuit that embeds data using the AngleEmbedding and then performs a variety of
+        operations. The output is a PauliZ measurement on the first output_dim qubits. One set of
+        parameters, w5, are specified as non-trainable."""
+        qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+        qml.templates.StronglyEntanglingLayers(w1, wires=list(range(n_qubits)))
+        qml.RX(w2, wires=0)
+        qml.RX(w3, wires=0)
+        qml.Rot(*w4, wires=0)
+        qml.templates.StronglyEntanglingLayers(w5, wires=list(range(n_qubits)))
+        return [qml.expval(qml.PauliZ(i)) for i in range(output_dim)]
+
+    return circuit, weight_shapes
+
+
+@pytest.mark.usefixtures("get_circuit")
+@pytest.fixture
+def model(get_circuit, n_qubits, output_dim):
+    """Fixture for creating a hybrid Keras model. The model is composed of KerasLayers sandwiched
+    between Dense layers."""
+    c, w = get_circuit
+    layer1 = KerasLayer(c, w, output_dim)
+    layer2 = KerasLayer(c, w, output_dim)
+
+    model = tf.keras.models.Sequential(
+        [
+            tf.keras.layers.Dense(n_qubits),
+            layer1,
+            tf.keras.layers.Dense(n_qubits),
+            layer2,
+            tf.keras.layers.Dense(output_dim),
+        ]
+    )
+
+    return model
+
+
+def indices(n_max):
+    """Returns an iterator over the number of qubits and output dimension, up to value n_max.
+    The output dimension never exceeds the number of qubits."""
+    a, b = np.tril_indices(n_max)
+    return zip(*[a + 1, b + 1])
+
+
+@min_tf
+@pytest.mark.usefixtures("get_circuit")
+class TestKerasLayer:
+    """Unit tests for the pennylane.qnn.KerasLayer class."""
+
+    @pytest.mark.parametrize("interface", ["tf"])
+    @pytest.mark.parametrize("n_qubits, output_dim", indices(1))
```

Comment:
Note: the new version of pytest that was just released (v5.4 I think) has 'fixed' the 'bug' where you could parametrize over implicitly used fixtures and not previously defined fixtures :( So you will need to either:

1. Add these parameters to the test function signatures, or

2. Re-write the fixtures to be factory functions